### ISSUE TITLE
Istio injector annotations don't work as described in doc

### ISSUE THREAD

**Issue by tcnghia** [istio/istio#6476]
**Describe the bug**
Istio injector annotations don't work as describe in at https://istio.io/docs/setup/kubernetes/sidecar-injection/#policy, 

* policy=disabled -> The sidecar injector will not inject the sidecar into pods by default. Add the sidecar.istio.io/inject annotation with value true to the pod template spec to enable injection.
* policy=enabled -> The sidecar injector will inject the sidecar into pods by default. Add the sidecar.istio.io/inject annotation with value false to the pod template spec to disable injection

**Expected behavior**
When policy=disabled the sidecar injector will not inject the sidecar into pods by default. Add the sidecar.istio.io/inject annotation with value true to the pod template spec to enable injection.

**Steps to reproduce the bug**
Create a pod in a namespace where policy=disabled or policy unset.  Injection will not happen no matter what we set pod annotation to.

**Version**
0.8

**Is Istio Auth enabled or not?**
No, auth not eabled.

```
helm template --namespace=istio-system \
    --set sidecarInjectorWebhook.enabled=true \
    --set global.proxy.image=proxyv2 \
    --set prometheus.enabled=false \
    install/kubernetes/helm/istio > ../istio.yaml
```
**Environment**
GKE 1.10

**Comment by tcnghia** [comment]
inject_test.yaml
```
apiVersion: v1
kind: Namespace
metadata:
  labels:
    istio-injection: enabled
  name: policy-enabled
---
apiVersion: v1
kind: Namespace
metadata:
  labels:
    istio-injection: disabled
  name: policy-disabled
---
apiVersion: v1
kind: Namespace
metadata:
  name: policy-default
---
apiVersion: v1
kind: Pod
metadata:
  name: nolabel
  namespace: policy-enabled
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
---
apiVersion: v1
kind: Pod
metadata:
  name: nolabel
  namespace: policy-disabled
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
---
apiVersion: v1
kind: Pod
metadata:
  name: nolabel
  namespace: policy-default
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
---
apiVersion: v1
kind: Pod
metadata:
  name: labeled-true
  namespace: policy-enabled
  annotations:
    "sidecar.istio.io/inject": "true"
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
---
apiVersion: v1
kind: Pod
metadata:
  name: labeled-true
  namespace: policy-disabled
  annotations:
    "sidecar.istio.io/inject": "true"
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
---
apiVersion: v1
kind: Pod
metadata:
  name: labeled-true
  namespace: policy-default
  annotations:
    "sidecar.istio.io/inject": "true"
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
---
apiVersion: v1
kind: Pod
metadata:
  name: labeled-false
  namespace: policy-enabled
  annotations:
    "sidecar.istio.io/inject": "false"
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
---
apiVersion: v1
kind: Pod
metadata:
  name: labeled-false
  namespace: policy-disabled
  annotations:
    "sidecar.istio.io/inject": "false"
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
---
apiVersion: v1
kind: Pod
metadata:
  name: labeled-false
  namespace: policy-default
  annotations:
    "sidecar.istio.io/inject": "false"
spec:
  containers:
  - image: docker.io/citizenstig/httpbin
    imagePullPolicy: IfNotPresent
    name: httpbin
```

```
$ kubectl get pods --all-namespaces=true | grep "label\|NAME"
NAMESPACE                NAME                                                READY     STATUS              RESTARTS   AGE
policy-default           labeled-false                                       1/1       Running             0          13s
policy-default           labeled-true                                        1/1       Running             0          13s
policy-default           nolabel                                             1/1       Running             0          13s
policy-disabled          labeled-false                                       1/1       Running             0          13s
policy-disabled          labeled-true                                        1/1       Running             0          13s
policy-disabled          nolabel                                             1/1       Running             0          13s
policy-enabled           labeled-false                                       1/1       Running             0          13s
policy-enabled           labeled-true                                        2/2       Running             0          13s
policy-enabled           nolabel                                             2/2       Running             0          13s
```

**Comment by prune998** [comment]
I'm not sure I understang your issue, but : 
- do a `kubectl get namespaces -o jsonpath='{.items[*].metadata.labels}'` to list namespaces and labels associated. You should see something like `map[istio-injection:enabled name:test]` if your namespace have the injectin enabled
- check the content of the configmap `istio-sidecar-injector` in your `istio-system namespace. You can use something like `kubectl -n istio-system get configmaps istio-sidecar-injector -o yaml`. Your config should start with `policy: disabled`
- check that your `deployment` includes the annotation `"sidecar.istio.io/inject": "false"` under the spec->template->metadata->annotations

Note that the Configmap is installed with `policy: enabled` by default so you have to use `--set global.proxy.policy=disabled` when using Helm

**Comment by ymesika** [comment]
@prune998, @tcnghia tested all combinations of having ns sidecar injector label vs having annotation in the pod.
His results are a bit confusing because he used the term `policy` for the NS label and `labeled` for the annotation.
The bottom line is that the label has high priority than the annotation. For instance, if you have a pod with an annotation to inject deployed into a ns with no label the pod won't get injected.

I asked him to open this issue following [his post here](https://groups.google.com/forum/#!topic/istio-users/y-ICsfgOWWY) because I think we need to clarify what's the expected behaviour for all combinations and fix/document it.

**Comment by tcnghia** [comment]
@ymesika the label doesn't always have higher priority than the annotation.  For example, when namespace policy label is `enabled` and pod annotation is `false` we see sidecar injection turned off.  So the annotation takes precedence in that case.  However, when policy label is `disabled`, it has higher priority.

The documentation https://istio.io/docs/setup/kubernetes/sidecar-injection/#policy explicitly states that 

> `policy=disabled` - The sidecar injector will not inject the sidecar into pods by default. Add the `sidecar.istio.io/inject` annotation with value `true` to the pod template spec to enable injection.

but this isn't the case as shown in the pod `policy-disabled/labeled-true` in my example.

@prune998  does this explain the issue?  thanks

I think with the current behavior it is hard to turn on sidecar injection for one particular pod.  We will need to opt-in the whole namespace, then opt-out all the other N-1 pod.  The behavior stated by the documentation at https://istio.io/docs/setup/kubernetes/sidecar-injection/#policy is better.

**Comment by tcnghia** [comment]
BTW, I saw that the documented policy is consistent with the code in https://github.com/istio/istio/blob/0759c8572119b8c78eeed1e0bbd3c58efc748092/pilot/pkg/kube/inject/inject.go#L277 

However, in my webhook logs the AdmissionReview didn't happen for Pods in a namespace with policy `disabled`.

I hope this information helps.

**Comment by ymesika** [comment]
@tcnghia I think you are confusing the policy from the doc with the namespace label. The policy in the doc refers to the `Policy` flag within the `istio-sidecar-injector` ConfigMap.

**Comment by tcnghia** [comment]
@ymesika I see now.  Can you please recommend how I use the opt-in policy described in the doc?  Do I set `policy=disabled` on ConfigMap and set `sidecar.istio.io/inject=true` on the Pod?

Thank

**Comment by prune998** [comment]
@tcnghia that's what I'm doing... No sidecar is injected except in the pods I want.
I think you still need to have `istio-injection:enabled` label on the `namespace`  for it to work

**Comment by tcnghia** [comment]
@ymesika It looks like setting up the ConfigMap policy works as you described.

I summarizes the behavior as followed.  

| Namespace label `istio-injection` | ConfigMap key `policy` | Pod annotation `sidecar.istio.io/inject` | Injection |
|-----------------------------------|------------------------|------------------------------------------|-----------|
| enabled                           | enabled                | true                                     | yes       |
| enabled                           | enabled                | false                                    | no        |
| enabled                           | disabled               | true                                     | yes       |
| enabled                           | disabled               | false                                    | no        |
| disabled                          | enabled                | true                                     | no        |
| disabled                          | enabled                | false                                    | no        |
| disabled                          | disabled               | true                                     | no        |
| disabled                          | disabled               | false                                    | no        |

tl;dr: if `namespace policy=disabled` then no injection.  Otherwise, if Pod annotation is `true` => injection, and Pod annotation is `false` => no injection.

In my case I have one namespace where I want to disable sidecar injection by default, except for one Pod (opt-in) (this namespace contains my controllers and webhooks).   I also have K namespaces where I want to enable sidecar injection by default (this namespace contains my users' containers).  Do you have recommendation of how I can achieve that?

Thanks a lot for the help so far.

**Comment by tcnghia** [comment]
If I understand correctly ConfigMap is a cluster-wide injection policy, namespace label `istio-injection` is a namespace level injection policy, and Pod annotation `sidecar.istio.io/inject` is a Pod level injection policy.

Since namespace level policy are not yet documented, may I request that we adopt a hierarchical model?   In such a hierarchical model, the policy with the least scope wins.  So `sidecar.istio.io/inject` has highest priority, followed by namespace label `istio-injection`, then injector config ConfigMap key `policy`.

That would simplify things and allow application frameworks to ensure sidecar injection on Pods that they created, no matter what namespace they are in.

Some more background: in our current use cases, we create Pods on behalf of our users, in their namespaces.  Our Pods should always get sidecar injection -- but currently we can't ensure that without asking the user to add the label to their namespace.

@ymesika @prune998 what do you think?

**Comment by ayj** [comment]
> tl;dr: if namespace policy=disabled then no injection. 

We don't have a namespace policy. The mutatingwebhookconfiguration's NamespaceSelector determines whether or not the webhook is invoked at all by the api-server. It can't be overridden by the per-pod annotation.

As an enhancement we could _replace_ the boolean `policy` option in the istio-inject ConfigMap with  namespace policy selector. 

1. mutatingwebhookconfiguration's NamespaceSelector => selects _when_ webhook is invoked
2. namespace policy selector => selects default opt-in / opt-out policy per namespace using label selector
3. annotation => allows pod-level overrides of namespace policy selector (#2).

**Comment by ostromart** [comment]
This is a documentation issue, dropping to P1.

**Comment by MartinKosicky** [comment]
I dont understand this, so how can I opt-in ?

**Comment by blankenshipz** [comment]
I'm surprised this hasn't been addressed; the documentation reads as though you can disable injection at the namespace and enable selectively in the pod template but that's apparently not the case?

**Comment by p1387h** [comment]
@blankenshipz: 
As far as i can tell, setting the "istio-injection: enabled" label at namespace level is required and there is no way around it, if you want to use the automatic injection. Selectively enabling a single pod does apparently not work. In my case i had to label the namesspace and manually blacklist every other pod via the ["neverjectSelector" entry](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#more-control-adding-exceptions) in the config map, which is far from optimal.

**Comment by johnking0099** [comment]
To summarize:

```
injected := namespace.labels["istio-injection"] == "enabled" &&
    ((configmap.policy == "enabled" && pod.annotations["sidecar.istio.io/inject"] != "false")
        || (configmap.policy == "disabled" && pod.annotations["sidecar.istio.io/inject"] == "true"))
```

**Comment by simonpie** [comment]
May I ask how we are supposed to follow the instructions for [tls prometheus integration ](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings) which specifically uses `sidecar.istio.io/inject: "true"`  ?  There are a bunch of pods in the prometheus' namespace, turning namespace injection would be a pain as we would need to retrofit a bunch of deployment to not deploy the sidecar.  

if I understand @johnking0099 last comments, I should go and modify my `configmap.policy` to `disabled` for this to work ?  But then, if I understand the table way [up there](https://github.com/istio/istio/issues/6476#issuecomment-399219937 ), it would break my current setup as I would have to specify the pod annotation everywhere.

This does not feel very user-friendly, and for sure, documentation is not as clear as it could be.

**Comment by boeboe** [comment]
I seems that I was not the only one totally confused here and losing 2 hours over it.

What I expected (and many people with me, it seems), is that if one does not annotate the namespace, but annotates the deployment/pod the enable injection, that injection would occur. This is the logical way and this is also how I read the documentation.

Why does Istio also seem to have a tendency to make things complicated and counter intuitive? The annotation closest the the target (pod is closer than namespace) should take precedence. It does, but it should NOT depend on a namespace annotation being set ALSO. I tried to look this weird constraint from all angles in the hope I missed some design decision, like safety by default or some other non-functional, but totally fail to find one.

@tcnghia thanks for the table, this is the only way to properly explain this, as opposed to all the fruitless ways of people here writing proza iso communicating clear.

**Comment by ruiengana** [comment]
According to https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#controlling-the-injection-policy

_The injector is configured with the following logic:

* If either label is disabled, the pod is not injected
* If either label is enabled, the pod is injected
* If neither label is set, the pod is injected if .values.sidecarInjectorWebhook.enableNamespacesByDefault is enabled. This is not enabled by default, so generally this means the pod is not injected._

PS: Labels above are referring to Namespace and Resource level labels.

With default `policy: enabled`, we can't validate above behaviour as sidecar isn't injected on a Pod with annotation `sidecar.istio.io/inject: "true"` running on a Namespace with label `istio-injection: disabled`.

**Comment by hemanthk76in** [comment]
Any updates on this? We need selective pod injection without associated namespace level injection. Is that possible?

**Comment by blankenshipz** [comment]
@hemanthk76in I don't think they're interested in making this work; I switched to linkerd for this usecase.

**Comment by boeboe** [comment]
@hemanthk76in @blankenshipz 

It is possible with the policy keymap set to disabled and namespace enabled. Only workloads with pod label enabled will be injected. Look at the full table above in this thread.

**Comment by liamhao** [comment]
> @boeboe 
> It is possible with the configmap key set to policy=disabled and namespace label istio-injection=enabled
------
IMHO, This is so anti-human 👎
